New C API <libdmmp/libdmmp.h>

cathay4t opened the pull request:
Features:
- Zero changes to existing codes.
- Utilizing existing libmultipath codes.
- Library user guide is in 'man 3 libdmmp.h'.
- Every public function has its own manpage in section 3.

Usage:

```
make -j5
sudo make install \
        bindir=/usr/sbin/ \
        syslibdir=/usr/lib64/ \
        libdir=/usr/lib64/multipath \
        rcdir=/etc/rc.d/init.d \
        unitdir=/usr/lib/systemd/system \
        includedir=/usr/include

man libdmmp.h
man dmmp_mpath_array_get
man <dmmp function name>
```

User case:
- Storaged multipath plugin. https://github.com/storaged-project/storaged/pull/40

FAQ:
1. Why not use better approach like wrapping multipathd IPC output?
   
   That often means a lot changes to existing code which might be
   rejected.
   I would like to create a stable set of API, while its internal
   implementation could be changed without breaking binary
   compatibility.
2. Why not build on existing libmultipath internal library?
   
   The libmultipath has too many public symbols which seems a bad
   design of public library. Yes, we still expose some internal symbols
   via libdmmp currently, that's because we are depending on
   libmultipath right now, to fix that we need to change libmultipath
   which I intend to avoid at this initial path set.
3. Any code style and rules?
   
   Following Linux kernel code style and libabc guideline.
   Other rules are recorded in 'libdmmp/DEV_NOTES'
4. Can the library be licensed as LGPL?
   
   Nope. LGPL library cannot link to any GPL code, while libmultipath
   is GPL. You could create some D-BUS API use libdmmp if license
   concerns. We might able to license libdmmp to LGPL when some day we
   change our implementation to wrap multipathd IPC output.
5. Why not expose all properties out?
   
   Let's do this step by step. This commit only contains minimum API
   set required to create the initial storaged multipath plugin.

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+		       uint32_t *mp_path_count)
+{
+	*mp_paths = NULL;
+	*mp_path_count = 0;
```
Missing { } after the if (based on the indentation)

cathay4t replied:
@trgill I am following [Linux kernel coding style](https://www.kernel.org/doc/Documentation/CodingStyle) here, queto:

```
Do not unnecessarily use braces where a single statement will do.

    if (condition)
        action();

and

    if (condition)
        do_this();
    else
        do_that();
```

trgill replied:
Sorry, I tagged the wrong line with this comment.  I was commenting about the indentation of this...   you don't want to indent the 2nd line since it isn't conditional, right?

if (mp_pg != NULL)
    *mp_paths = mp_pg->mp_paths,
    *mp_path_count = mp_pg->mp_path_count;  // this line shouldn't be indented or it should be surrounded with {}

cathay4t replied:
Thanks for pointing out.
Fixed.

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+	if (mp_pg != NULL)
+		*mp_paths = mp_pg->mp_paths,
+		*mp_path_count = mp_pg->mp_path_count;
+	return 0;
```
Should the function just return void rather than always 0?

cathay4t replied:
Yeah. I will change it to void.

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+		       struct dmmp_path ***mp_paths,
+		       uint32_t *mp_path_count)
+{
+	*mp_paths = NULL;
```
check to make sure these are not NULL before dereferencing?

cathay4t replied:
I don't get it. Which line you are referring?

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+	if (mpath != NULL)
+		*mp_pgs = mpath->mp_pgs;
+		*mp_pg_count = mpath->mp_pg_count;
+	return 0;
```
Maybe you are planning on returning an error in the case mp_pgs || mp_pg_count are NULL?  If not change to return void?

cathay4t replied:
void will be.

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+		if (mpath == NULL)
+			goto error_out;
+
+		*mpaths[i] = mpath;
```
this line gets a SIGSEGV on a machine with real multipath devices

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+		goto error_out;
+
+	*mpaths = (struct dmmp_mpath **)
+		MALLOC(sizeof(struct dmmp_mpath *) * VECTOR_SIZE(mppvec));
```
I think you need a path count here.  VECTOR_SIZE returns the number maps, right?  If you've got more than 1 path per map you won't have enough space.

cathay4t replied:
The `mpath->mp_pg_count` and `mpath->mp_pgs` are holding the path groups.
The paths are saved in `struct dmmp_path_group` as:
`mp_pg->mp_paths` and `mp_pg->mp_path_count`.

trgill replied:
Sorry, my comment isn't clear.

When I run your test program on a system with 7 maps and 14 paths, I get a SEGV and path_count is set to 7.  I'll email you a test system so you can have a look.

cathay4t replied:
Thanks. It was caused by:

```
diff --git a/libmultipath/public_api.c b/libmultipath/public_api.c
index dff05fb..4429220 100644
--- a/libmultipath/public_api.c
+++ b/libmultipath/public_api.c
@@ -487,7 +487,7 @@ int dmmp_mpath_list(struct dmmp_mpath ***mpaths, uint32_t *mpath_count)
                if (mpath == NULL)
                        goto error_out;

-               *mpaths[i] = mpath;
+               (*mpaths)[i] = mpath;

                if (mpp == NULL)
                        goto error_out;
```

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+	int rc = 0;
+	int i = 0;
+
+	*mpaths = NULL;
```
NULL checks?

cathay4t replied:
Sorry, which line you are referring?

trgill replied:
If the mpaths parameter is NULL, *mpaths will SEGV right?  This is minor, but why not check to make sure it is non-NULL?

cathay4t replied:
That's a good point. I will fix them.
Thanks.

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+	struct path * pp;
+
+	if (!mpp->pg)
+		return 0;
```
error case? return something other than 0?

cathay4t replied:
This method is copy from `multipath/main.c update_paths()`. I would like to keep it in
original form.

trgill commented on libmultipath/public_api.c:
```diff
@@ -0,0 +1,794 @@
+	strcpy(mp_path->name, pp->dev);
+	if (pp->state > 0)
+		mp_path->status = pp->state;
+	return 0;
```
There are a few functions that return only zero.  If there is no error case, maybe consider making the function return void?

cathay4t replied:
void will be.
Thanks.

trgill commented on libdmmp/libdmmp.c:
```diff
@@ -0,0 +1,437 @@
+	dm_lib_release();
+	dm_lib_exit();
+
+	cleanup_prio();
```
Is this necessary?  The code doesn't initialize the the checkers/prioritizers, right?  Do we need to call the cleanup_xxx calls?

cathay4t replied:
Opps, I forgot to remove this part.
Thanks.

trgill commented on libdmmp/libdmmp.c:
```diff
@@ -0,0 +1,437 @@
+	int rc_discovery = 0;
+
+	*pathvec = vector_alloc();
+	/* No need to whether pathvec is NULL, caller already done */
```
*pathvec can't be NULL, but vector_alloc() could return NULL.  Shouldn't we check that?

cathay4t replied:
The `_dmmp_alloc_null_check(ctx, *pathvec, rc, out);` will check whether `*pathvec` is NULL or not.

trgill commented on libdmmp/libdmmp_misc.c:
```diff
@@ -0,0 +1,84 @@
+
+static const char * const _DMMP_RC_MSG[] = {
+	"OK",
+	"Ran out of memory",
```
I would suggest to remove the word "Ran" and change to:

"Out of memory"

cathay4t replied:
Changed to "Out of memory".

trgill commented on libdmmp/libdmmp_pg.c:
```diff
@@ -0,0 +1,175 @@
+	free(dmmp_pg);
+}
+
+void dmmp_path_array_get(struct dmmp_path_group *mp_pg,
```
I think you intended to put the path_group functions in this file, right?  This "path" function should be in the libdmmp_path.c file?

cathay4t replied:
No. The `dmmp_path_array_get()` functions is accessing properties (`dmmp_ps` and `dmmp_p_count`) of `struct dmmp_path_group` which is defined in `libdmmp_pg.c`, so it has to stay in `libdmmp_pg.c`.

trgill commented on libdmmp/libdmmp_mp.c:
```diff
@@ -0,0 +1,163 @@
+	free(dmmp_mp);
+}
+
+void dmmp_path_group_array_get(struct dmmp_mpath *dmmp_mp,
```
Does  dmmp_path_group_array_get belong in libdmmp_pg.c?

cathay4t replied:
No. The same reason as `dmmp_path_array_get()` function.

cathay4t committed:
Introducing multipath C API <libdmmp/libdmmp.h>

Features:

 * Use three multipathd IPC commands:
    * 'show maps raw format <fmt>'
    * 'show groups raw format <fmt>'
    * 'show paths raw format <fmt>'
    # We might have inconsistent data when something changes during these
    # commands. If so, DMMP_ERR_INCONSISTENT_DATA will be raised and
    # suggest user to try again.
 * Library user guide will be 'man 3 libdmmp.h'.
 * Every public function has its own manpage in section 3
   generated by linux 'kernel-doc' tool.

Usage:

    make -j5
    sudo make install \
            bindir=/usr/sbin/ \
            syslibdir=/usr/lib64/ \
            libdir=/usr/lib64/multipath \
            rcdir=/etc/rc.d/init.d \
            unitdir=/usr/lib/systemd/system \
            includedir=/usr/include
    make -C libdmmp check

    man libdmmp.h
    man dmmp_mpath_array_get
    man <dmmp function name>

Performance:

  * 4k scsi_debug sdX with 2 disks per mpath (i7-3520M 4GiB RAM RHEL 7.2):
        $ make -C libdmmp speed_test
        Got 2000 mpath
        real 0.20
        user 0.03
        sys 0.01
  * 10k scsi_debug sdX with 2 disks per mpath (E5-2697 32GiB RAM RHEL 7.2):
        $ make -C libdmmp speed_test
        Got 5000 mpath
        real 1.51
        user 0.45
        sys 0.00

User case:

    Storaged multipath plugin:
        https://github.com/storaged-project/storaged/pull/40

FAQ:

 1. Any developer notes?

    Following Linux kernel code style and libabc guideline.
    Others are recorded in 'libdmmp/DEV_NOTES'

 2. Can the library be licensed as LGPL?

    Nope. I copied some socket codes from libmultipath which is GPL.
    You could create some D-BUS API use libdmmp if license concerns.
    We might able to license libdmmp to LGPL after rewrite those functions.

Signed-off-by: Gris Ge <fge@redhat.com>

trgill committed:
Add %g format specifier to 'multipathd show paths format' command

The %g specifier will print the group number of the path when used with
multipathd show paths format.

Signed-off-by: Todd Gill <tgill@redhat.com>
Signed-off-by: Gris Ge <fge@redhat.com>

trgill committed:
Add 'show groups' command to multipathd

Command allows user to display pathgroups.  This is a step towards
allowing a user to parse the output of multipathd and be able to
understand the groups.

Changes by Gris: (changes since v1)
 * Remove 'update_multipath()' in show_groups() as it only update path
   status which is not required for path group query.
 * Disable white space padding in raw mode.
 * Display first group as 1 not 0

Signed-off-by: Todd Gill <tgill@redhat.com>
Signed-off-by: Gris Ge <fge@redhat.com>

trgill committed:
Add options to the show groups command

Options:

format - allow the user to specify output format via wildcards
raw    - don't display headers for formatted output
%w     - display the WWN for the map that contains the group
%g     - display the group number

These options allow a user to parse the output to connect a pathgroup
to the multipath map.  It also allows a user to determine the member
paths of a pathgroup.

Signed-off-by: Todd Gill <tgill@redhat.com>
Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
Closing this PR.
Patch has been send to upstream mail list( dm-devel@redhat.com) for review.
Please send feedback to there.
Thank  you.
